features/events/daemon/src/main/java/org/opennms/netmgt/eventd/processor/expandable/EventTemplate.java

```diff
@@ -0,0 +1,137 @@
+
+    private final String input;
+
+    private final List<ExpandableToken> tokens = Lists.newArrayList();
+
+    private final EventUtil eventUtil;
+
+    private final ExpandableParameterResolverRegistry parameterResolverRegistry;
+
+    public EventTemplate(String input, EventUtil eventUtil) {
+        this.input = Objects.requireNonNull(input);
+        this.eventUtil = Objects.requireNonNull(eventUtil);
+        this.parameterResolverRegistry = new ExpandableParameterResolverRegistry();
+        parse();
+    }
+
+    /**
+     * Parses the input and creates {@link ExpandableToken} to expand it.
+     */
+    private void parse() {
+        tokens.clear();
+        String tempInp = input;
+        int inpLen = input.length();
+
+        int index1 = -1;
+        int index2 = -1;
+
+        // check input string to see if it has any %xxx% substring
+        while ((tempInp != null) && ((index1 = tempInp.indexOf(PERCENT)) != -1)) {
+
+            LOG.debug("checking input {}", tempInp);
+            // copy till first %
+            tokens.add(new ExpandableConstant(tempInp.substring(0, index1)));
+            tempInp = tempInp.substring(index1);
+
+            index2 = tempInp.indexOf(PERCENT, 1);
+            if (index2 != -1) {
+                // Get the value between the %s
+                String parm = tempInp.substring(1, index2);
+                LOG.debug("parm: {} found in value", parm);
+
+                // If there's any whitespace in between the % signs, then do not try to
+                // expand it with a parameter value
+                if (parm.matches(WHITESPACE_PATTERN)) {
+                    tokens.add(new ExpandableConstant(PERCENT));
+                    tempInp = tempInp.substring(1);
+                    LOG.debug("skipping parm: {} because whitespace found in value", parm);
+                    continue;
+                }
+
+                tokens.add(new ExpandableParameter(parm, parameterResolverRegistry.getResolver(parm), eventUtil));
+
+                if (index2 < (inpLen - 1)) {
+                    tempInp = tempInp.substring(index2 + 1);
+                } else {
+                    tempInp = null;
+                }
+            } else {
+                break;
+            }
+        }
+        if ((index1 == -1 || index2 == -1) && (tempInp != null)) {
+            tokens.add(new ExpandableConstant(tempInp));
+        }
+    }
+
+    @Override
+    public String expand(Event event, Map<String, Map<String, String>> decode) {
+        final String collect = tokens.stream()
+                .map(t -> t.expand(event, decode))
+                .collect(Collectors.joining());
+        return collect;
+    }
+
+    // If we find any token which requires a transaction, the template itself requires a transaction as well
+    @Override
+    public boolean requiresTransaction() {
+        return tokens.stream().filter(t -> t.requiresTransaction()).findAny().isPresent();
```

Comment:
Can we pre-compute this value instead? The `tokens` are immutable and set in the constructor.